pkg/ingester/client/cortex.proto

```diff
@@ -13,6 +13,7 @@ service Ingester {
   rpc Push(WriteRequest) returns (WriteResponse) {};
   rpc Query(QueryRequest) returns (QueryResponse) {};
   rpc QueryStream(QueryRequest) returns (stream QueryStreamResponse) {};
+  rpc QueryTSDBStream(QueryRequest) returns (stream QueryTSDBStreamResponse) {};
```

Comment:
I'm a bit dubious about this duplication, instead of using `QueryStream`. Is it hard to fit the TSDB response into `QueryStreamResponse`? If I'm not missing anything, it would allow use to reuse the code on the distributor side, instead of the custom `ingesterTSDBQueryable`.

Reply:
So my reasoning was it didn't make sense to re-purpose it, even though that would have been ideal. Since a `QueryStreamResponse` returns a slice of `TimeSeriesChunk` and the TSDB select returns a `SeriesSet` it didn't make sense to me to then re-encode that data into chunks just to reuse a function.

Reply:
I see your point. However, I think it's not just a function reusing, but the large amount of the new code introduced in this PR. Am I wrong?

Reply:
I don't think fear of new code is a good reason to make a bad design decision. If it's "better" to re-encode into chunks before streaming then it would be worth it

Reply:
I don't fear new code, but I believe when we - as software engineers - evaluate options, we should balance different things, including whether it's worth or not to fit in the current design.

Given we're in the discussion phase (when there's no design doc, the code review is also the place where tradeoffs in design are evaluated), I would be glad if we could openly evaluate pros and cons.

That being said, I do understand your perspective, but I've the feeling that re-using Cortex chunks is - from a design perspective - what we already did in `BlockQuerier.Get()` when we call the `seriesToChunks()`?

Reply:
I'm not arguing we shouldn't have this conversation here :) we definitely should. 

I definitely wanted to re-use as much code as possible, but since nothing else in the TSDB path operates in chunks it felt a diversion from that, as well as a waste of CPU cycles to encode it back to chunks.

Reply:
I'm not against your perspective. Let me just understand, before I do another round of review. Would be different than what we already did in BlockQuerier.Get() when we call the seriesToChunks() or would be the same? _Just to understand, maybe I'm missing something._

Reply:
It would be the same in concept, but different in implementation. `seriesToChunks` is converting one set of chunks to a different type of chunks. We'd have to write a timeseriesToChunks function to convert the timeseries slice. 

The block querier already deals with chunks, so it felt like less of a stretch from a high-level perspective.

Reply:
Ingester with chunks simply sends entire encoded chunks to the querier, which is interesting design choice, most likely done to avoid doing extra work inside ingester.

We obviously cannot do the same with TSDB, since its API is different. We need to extract data from TSDB, but encoding to chunks is extra work – and we try to put extra work in queriers instead. I think a good compromise would be to unify the API and return both results in `QueryStreamResponse` (in different fields), and let querier handle those differences.

This way, querier doesn't need to know whether ingester uses chunks of TSDB beforehand (to choose API call), but will simply work with whatever data ingester returns (chunks or timeseries).

Reply:
@pstibrany that's a good idea. I've implemented a first pass of that idea. API now returns a `QueryStreamResponse` which contains fields for a slice of Timeseries or Chunkseries. 

I considered leveraging grpc using the oneof field
```
message QueryStreamResponse {
  repeated QueryStreamData series = 1;
}

message QueryStreamData {
  oneof type {
    TimeSeriesChunk chunkseries = 1 [(gogoproto.nullable) = false];
    TimeSeries timeseries = 2 [(gogoproto.nullable) = false];
  }
}
```
which has the benefit it guarantees only one of those structs can exist in the response, but I felt like it made the code more obtuse. Thoughts?

Reply:
I cannot see generated code when using `oneOf`, so cannot comment on complexity, but I wouldn't worry about `oneOf` too much. Putting a comment into proto file explaining the contract (only one of the fields is set) would be enough imho, as this is internal API.

Reply:
I've found this: https://developers.google.com/protocol-buffers/docs/reference/go-generated#oneof, it doesn't look *that* terrible :)

Reply:
alright, talked me into it :)

Reply:
Nevermind, it quickly became a mess getting it to work with our existing code. Leaving as is for now.